ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2151,6 +2159,173 @@ prop_connection_manager_valid_transition_order serverAcc (ArbDataFlow dataFlow)
                        maxAcceptedConnectionsLimit l
 
+-- | Property wrapping `multinodeExperiment`.
+--
+-- Note: this test validates connection manager counters using an upper bound
+-- approach since there's no reliable way to reconstruct the value that the
+-- counters should have at a given point in time. It's not quite possible to
+-- have the god's view of the system consistent with the information that's
+-- traced because there can be timing issues where a connection is in
+-- TerminatingSt (hence counted as 0) but in the God's view the connection is
+-- still being deleted (hence counted as 1). This is all due to the not having
+-- a better way of injecting god's view traces in a way that the timing issues
+-- aren't an issue.
+--
+prop_connection_manager_counters :: Int
+                                 -> ArbDataFlow
+                                 -> MultiNodeScript Int TestAddr
+                                 -> Property
+prop_connection_manager_counters serverAcc (ArbDataFlow dataFlow)
+                                 script@(MultiNodeScript l) =
+  let trace = runSimTrace sim
+
+      evsCMT :: Trace (SimResult ())
+                        (ConnectionManagerTrace
+                          SimAddr
+                          (ConnectionHandlerTrace
+                            UnversionedProtocol
+                            DataFlowProtocolData))
+      evsCMT = traceWithNameTraceEvents trace
+
+      -- Needed for calculating a more accurate upper bound
+      evsUS :: [ObservableNetworkState SimAddr]
+      evsUS = selectTraceEventsDynamic trace
+
+      upperBound =
+        collapseCounters True
+        $ multiNodeScriptToCounters dataFlow l evsUS
+
+  in tabulate "ConnectionEvents" (map showCEvs l)
+    . counterexample (ppScript script)
+    . counterexample (Trace.ppTrace show show evsCMT)
+    . counterexample (intercalate "\n" $ map show evsUS)
+    . getAllProperty
+    . bifoldMap
+       ( \ case
+           MainReturn {} -> mempty
+           v             -> AllProperty
+                            $ counterexample (show v) (property False)
+       )
+       ( \ trs
+        -> case trs of
+          TrConnectionManagerCounters cmc ->
+            AllProperty
+              $ counterexample
+                  ("Upper bound is: " ++ show upperBound
+                  ++ "\n But got: " ++ show cmc)
+                  (property $ collapseCounters False cmc <= upperBound)
+          _                               ->
+            mempty
+       )
+    $ evsCMT
+  where
+    serverAddress :: SimAddr
+    serverAddress = TestAddress 0
+
+    -- We count all connections as prunable because we do not have a better way
+    -- to know what transitions will a given connection go through. We also
+    -- count every client connection as unidirectional because they do not
+    -- negotiate duplex data flow.
+    --
+    -- We do not count Incoming/Outgoing/Closing of connections since the
```

Comment:
just for the consistency sake, we call them inbound / outbound rather than incoming / outgoing.  I'd like to change this in counters too at some stage.